### PR TITLE
Add support for 2FA code in Apple logins when requested

### PR DESCRIPTION
**Pull request by squallstar**
### Login flow

image

### Load teams

Teams will no longer be automatically loaded when you open the AAB UI. This is because 2FA might be required, and if it's automated process, users will be unexpectedly asked to enter 2FA code.

They don't need to enter login details again, though. We've simply added a **Load teams** button, which will load the teams on demand -- and might require 2FA verification using the UI above if necessary.

image

### CODE REVIEW

File: js/interface.js [hunk 1]
```diff
@@ -2821,4 +2822,41 @@ function initialLoad(initial, timeout) {
...
+  // Ask user for code
+  prompt2FACode().then(function (code) {
+    if (code === null) {
+      // @TODO Cancels login
+      return;
+    }
+
+    // Notify requester with the code
+    // @Question What happens when 2FA fails?
+    socket.to(data.clientId).emit('aab.apple.login.2fa.code', code);
```
Comment (tonytlwu) [inline js/interface.js]: @squallstar How can I detect and handle 2FA login failure?

Comment (tonytlwu) [reply js/interface.js]: ...or success, for that matter.

Comment (squallstar) [reply js/interface.js]: @tonytlwu I'll see if I can add two events for you to listen, like `aab.apple.login.2fa.success` and `aab.apple.login.2fa.failure`

File: js/interface.js [hunk 2]
```diff
@@ -2821,4 +2822,41 @@ function initialLoad(initial, timeout) {
...
+  return Fliplet.Modal.prompt({
+    title: 'Please enter the verification code to verify your login'
+  }).then(function (code) {
+    if (code === null) {
+      return null;
+    }
+
+    if (code === '') {
+      return Fliplet.Modal.alert({
+        message: 'You must enter the verifiaction code to log in'
```
Comment (squallstar) [inline js/interface.js]: typo in `verifiaction`

File: js/interface.js [hunk 3]
```diff
@@ -2821,4 +2822,41 @@ function initialLoad(initial, timeout) {
...
+    return code;
+  });
+}
+
+// Listen for 2FA code when requested
+socket.on('aab.apple.login.2fa', function (data) {
+  // Ask user for code
+  prompt2FACode().then(function (code) {
+    if (code === null) {
+      // @TODO Cancels login
```
Comment (squallstar) [inline js/interface.js]: add this: `socket.to(data.clientId).emit('aab.apple.login.2fa.cancel');`

File: js/interface.js [hunk 4]
```diff
@@ -1967,8 +1967,38 @@ function initialLoad(initial, timeout) {
...
+  }).then(function (selection) {
+    if (selection === null) {
+      return null;
+    }
+
+    if (selection === '') {
+      return Fliplet.Modal.alert({
+        message: 'You must select a device to verify your identity'
+      }).then(function () {
+        return select2FAMethod();
```
Comment (squallstar) [inline js/interface.js]: `select2FAMethod(data);`

**Comment by tonytlwu** [comment]
@squallstar I moved chunks of functions (no refactoring) around so that they're all sitting in the same place in the code. Because of this your changes made to `master` is creating a pretty complex conflict. Could you take a look and resolve it?

File: js/interface.js [hunk 5]
```diff
@@ -1509,1340 +1582,1494 @@ function validateScreenshots() {
...
+  if (devEmail === '') {
+    $('#fl-store-appDevLogin').parents('.form-group').addClass('has-error has-danger');
+    $('#fl-store-appDevLogin').next('.with-errors').html(emailError);
+    $(this).html('Log in');
+    $(this).removeClass('disabled');
+    Fliplet.Widget.autosize();
   }
-}
-
-
```
Comment (squallstar) [inline js/interface.js]: what's this for?

Comment (tonytlwu) [reply js/interface.js]: I needed some personal space. Please, just back off.

:trollface:

**Pull request merged**